lib/bankscrap/ing/bank.rb

```diff
@@ -140,64 +139,58 @@ def post_auth(ticket)
-          current_pinpad_paths << tmp.path
+      def extract_images_from_pinpad(pinpad)
+        pinpad.map do |string|
+          extract_info_from_encoded_png(string)
         end
-
-        current_pinpad_paths
       end
 
-      def recognize_pinpad_numbers(current_pinpad_paths)
-        real_numbers = []
-        current_numbers = Magick::ImageList.new(*current_pinpad_paths)
+      def extract_info_from_encoded_png(encoded_png)
+        decoded = Base64.decode64(encoded_png)
+        # PNG format first bytes are headers that we can discard. We know where the data chunk starts,
+        # so we can directly read the image data. See: https://www.w3.org/TR/PNG-Structure.html
+        length  = decoded.slice(33, 4).unpack('L>').first
+        data    = decoded.slice(41, length)
+        Zlib::Inflate.inflate(data)       # We decompress the data
+                     .delete("\u0000")    # Levenshtein distance doesn't work with null bytes
+                     .slice(2_000, 5_000) # The whole image is too slow to compare, we can use a chunk in the middle.
+      end
+
+      def recognize_pinpad_numbers(received_pinpad)
+        # For each image received, we compare it with the images of the numbers we already know,
+        # looking for the most similar one (using Levenshtein's string comparison)
+        received_pinpad_images = extract_images_from_pinpad(received_pinpad)
+        sorted_pinpad_images = extract_images_from_pinpad(Config::SORTED_PINPAD)
+        received_pinpad_numbers = []
+        candidates = (0..9).to_a
         0.upto(9) do |i|
-          pixel_diffs = []
-          0.upto(9) do |j|
-            sample_number_img = Magick::ImageList.new(sample_number_path(j)).first
-            pixel_diffs << images_diff(sample_number_img, current_numbers[i])
-          end
-          real_numbers << pixel_diffs.each_with_index.min.last
+          number = identify_more_similar_image(sorted_pinpad_images, received_pinpad_images[i], candidates)
+          candidates.delete(number) # We can reduce our search space
+          received_pinpad_numbers << number
         end
-        real_numbers
-      end
-
-      def sample_number_path(number)
-        File.join(File.dirname(__FILE__), "#{SAMPLE_ROOT_PATH}/#{number}.png")
-      end
-
-      def images_diff(sample_number_img, current_number_img)
-        diff = 0
-        sample_pixels  = pixels_from_coordinates(sample_number_img, 0, 0)
-        current_pixels = pixels_from_coordinates(current_number_img, 0, 0)
-        current_pixels.each_with_index do |pixel, index|
-          sample_pixel = sample_pixels[index]
-          red_diff = (pixel.red - sample_pixel.red).abs
-          green_diff = (pixel.green - sample_pixel.green).abs
-          blue_diff = (pixel.blue - sample_pixel.blue).abs
-
-          diff += red_diff + green_diff + blue_diff
+        received_pinpad_numbers
+      end
+
+      def identify_more_similar_image(sorted_pinpad_images, image, candidates)
```

Comment:
Maybe `find_most_similar_image` could be more readable and grammatically correct? Just sayin', if you prefer `identify_more_similar-image` is ok.